finality-aleph/src/network/substrate.rs

```diff
@@ -60,163 +83,230 @@ impl ProtocolNaming {
     }
 }
 
-/// A struct holding NotificationService per protocol we use.
-pub struct NotificationServices {
-    pub authentication: Box<dyn sc_network::config::NotificationService>,
-    pub sync: Box<dyn sc_network::config::NotificationService>,
+#[derive(Debug)]
+pub enum SyncNetworkServiceError {
+    NetworkStreamTerminated,
+}
+
+impl fmt::Display for SyncNetworkServiceError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::NetworkStreamTerminated => write!(f, "Network event stream ended."),
+        }
+    }
 }
 
-pub struct NetworkEventStream<B: Block, H: ExHashT> {
+/// Service responsible for handling network events emitted by the base sync protocol.
+pub struct SyncNetworkService<B: Block, H: ExHashT> {
     sync_stream: Fuse<Pin<Box<dyn Stream<Item = SyncEvent> + Send>>>,
-    notifications: NotificationServices,
     naming: ProtocolNaming,
     network: Arc<NetworkService<B, H>>,
 }
 
-impl<B: Block, H: ExHashT> NetworkEventStream<B, H> {
+impl<B: Block, H: ExHashT> SyncNetworkService<B, H> {
     pub fn new(
         network: Arc<NetworkService<B, H>>,
         sync_network: Arc<SyncingService<B>>,
         naming: ProtocolNaming,
-        notifications: NotificationServices,
     ) -> Self {
         Self {
             sync_stream: sync_network.event_stream("aleph-syncing-network").fuse(),
-            notifications,
             naming,
             network,
         }
     }
-}
 
-#[async_trait]
-impl<B: Block, H: ExHashT> EventStream<PeerId> for NetworkEventStream<B, H> {
-    async fn next_event(&mut self) -> Option<Event<PeerId>> {
-        use Event::*;
+    fn protocols(&self) -> [Protocol; 2] {
+        [Protocol::Authentication, Protocol::BlockSync]
+    }
+
+    fn peer_connected(&mut self, remote: PeerId) {
+        let multiaddress: Multiaddr =
+            iter::once(MultiaddressProtocol::P2p(remote.into())).collect();
+        trace!(target: LOG_TARGET, "Connected event from address {:?}", multiaddress);
+
+        for protocol in self.protocols() {
+            let name = self.naming.protocol_name(&protocol);
+            if let Err(e) = self
+                .network
+                .add_peers_to_reserved_set(name.clone(), iter::once(multiaddress.clone()).collect())
+            {
+                error!(target: LOG_TARGET, "add_reserved failed for {}: {}", name, e);
+            }
+        }
+    }
+
+    fn peer_disconnected(&mut self, remote: PeerId) {
+        trace!(target: LOG_TARGET, "Disconnected event for peer {:?}", remote);
+        let addresses: Vec<_> = iter::once(remote).collect();
+
+        for protocol in self.protocols() {
+            let name = self.naming.protocol_name(&protocol);
+            if let Err(e) = self
+                .network
+                .remove_peers_from_reserved_set(name.clone(), addresses.clone())
+            {
+                warn!(target: LOG_TARGET, "Error while removing peer from {} reserved set: {}", name, e)
```

Comment:
in peer connected this is error, here its warn - any reason for tht?

Reply:
you tell me...

Reply:
lets make it error for consistency :P

Reply:
done